File: src/main/java/wiresegal/psionup/common/items/ItemGaussRifle.java
package wiresegal.psionup.common.items;

import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.item.ItemStack;
import net.minecraft.network.play.server.SPacketEntityVelocity;
import net.minecraft.util.*;
import net.minecraft.world.World;
import vazkii.psi.api.PsiAPI;
import vazkii.psi.api.cad.ICAD;
import vazkii.psi.common.core.handler.PlayerDataHandler;
import vazkii.psi.common.core.handler.PsiSoundHandler;
import wiresegal.psionup.common.core.helper.flowcolors.IFlowColorAcceptor;
import wiresegal.psionup.common.entity.EntityGaussPulse;

import static wiresegal.psionup.common.entity.EntityGaussPulse.AmmoStatus.*;

public class ItemGaussRifle extends PsionicItem implements IFlowColorAcceptor {
	public ItemGaussRifle(ResourceLocation res) {
		super(res);
		
		setMaxStackSize(1);
	}
	
	private static final int PSI_COST_NO_AMMO = 625;
	private static final int PSI_COST_WITH_AMMO = 200;
	
	//TODO this liblib glowing item shit, reimplement that somehow
	//TODO iitemcolor
	
	@Override
	public ActionResult<ItemStack> onItemRightClick(World world, EntityPlayer player, EnumHand hand) {
		PlayerDataHandler.PlayerData data = PlayerDataHandler.get(player);
		ItemStack cad = PsiAPI.getPlayerCAD(player);
		if(cad.isEmpty()) return new ActionResult<>(EnumActionResult.SUCCESS, player.getHeldItem(hand));
		
		ItemStack ammo = findAmmo(player);
		boolean creative = player.capabilities.isCreativeMode;
		
		if(creative || data.availablePsi > 0) {
			boolean usingAmmo = !ammo.isEmpty();
			boolean bottomedOutPsi = false;
			
			//Drain psi or consume ammo item
			if(!creative) {
				if(usingAmmo) {
					data.deductPsi(PSI_COST_WITH_AMMO, 10, true);
					ammo.shrink(1);
				} else {
					int cadBattery = ((ICAD)cad.getItem()).getStoredPsi(cad);
					if(data.availablePsi + cadBattery < PSI_COST_NO_AMMO) bottomedOutPsi = true;
					data.deductPsi(PSI_COST_NO_AMMO, (int) (3 * player.getCooldownPeriod()) + (bottomedOutPsi ? 60 : 10), true);
				}
			}
			
			player.swingArm(hand);
			
			EntityGaussPulse.AmmoStatus status;
			
			if(usingAmmo) status = creative ? DEPLETED : AMMO;
			else status = bottomedOutPsi ? BLOOD : NOTAMMO;
			
			EntityGaussPulse ent = new EntityGaussPulse(world, player, status);
			
			player.motionX -= .50 * player.getLookVec().x;
			player.motionY -= .25 * player.getLookVec().y;
			player.motionZ -= .50 * player.getLookVec().z;
			
			if(!world.isRemote) {
				world.spawnEntity(ent);
				
				((EntityPlayerMP) player).getServerWorld().getEntityTracker().sendToTrackingAndSelf(player, new SPacketEntityVelocity(player));
			}
			
			world.playSound(null, player.posX, player.posY, player.posZ, status == BLOOD ? PsiSoundHandler.compileError : PsiSoundHandler.cadShoot, SoundCategory.PLAYERS, 1f, 1f);
			
			if(!creative) {
				player.getCooldownTracker().setCooldown(this, (int) player.getCooldownPeriod() * 3);
			}
		}
		
		return new ActionResult<>(EnumActionResult.SUCCESS, player.getHeldItem(hand));
	}
	
	private ItemStack findAmmo(EntityPlayer player) {
		//Prioritize held ammo over the rest of the inventory (like a bow)
		
		for(EnumHand hand : EnumHand.values()) {
			ItemStack held = player.getHeldItem(hand);
			if(isAmmo(held)) return held;
		}
		
		for(int slot = 0; slot < player.inventory.getSizeInventory(); slot++) {
			ItemStack stackInSlot = player.inventory.getStackInSlot(slot);
			if(isAmmo(stackInSlot)) return stackInSlot;
		}
		
		return ItemStack.EMPTY;
	}
	
	private boolean isAmmo(ItemStack stack) {
		return !stack.isEmpty() && stack.getItem() == ModItems.gaussBullet;
	}
}

File: src/main/java/wiresegal/psionup/common/items/ItemGaussRifle.kt
package wiresegal.psionup.common.items

import net.minecraft.client.renderer.color.IItemColor
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.util.ActionResult
import net.minecraft.util.EnumActionResult
import net.minecraft.util.EnumHand
import net.minecraft.util.SoundCategory
import net.minecraft.util.math.MathHelper
import net.minecraft.world.World
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import vazkii.psi.api.cad.ICADColorizer
import vazkii.psi.client.core.handler.ClientTickHandler
import vazkii.psi.common.core.handler.PlayerDataHandler
import vazkii.psi.common.core.handler.PsiSoundHandler
import wiresegal.psionup.client.core.handler.ModelHandler
import wiresegal.psionup.common.entity.EntityGaussPulse
import wiresegal.psionup.common.items.base.ItemMod

/**
 * @author WireSegal
 * Created at 10:10 PM on 7/13/16.
 */
class ItemGaussRifle(name: String) : ItemMod(name), ModelHandler.IItemColorProvider {
    override fun onItemRightClick(itemStackIn: ItemStack, worldIn: World, playerIn: EntityPlayer, hand: EnumHand): ActionResult<ItemStack> {
        val data = PlayerDataHandler.get(playerIn)
        val ammo = findAmmo(playerIn)
        if (ammo != null || data.availablePsi > 0 || playerIn.capabilities.isCreativeMode) {
            if (!playerIn.capabilities.isCreativeMode) {
                if (ammo == null)
                    data.deductPsi(1000, 100, true)
                else {
                    ammo.stackSize--
                    if (ammo.stackSize == 0)
                        playerIn.inventory.deleteStack(ammo)
                }
            }

            playerIn.swingArm(hand)

            val status = if (ammo != null) {
                if (playerIn.capabilities.isCreativeMode)
                    EntityGaussPulse.AmmoStatus.DEPLETED
                else
                    EntityGaussPulse.AmmoStatus.AMMO
            } else
                EntityGaussPulse.AmmoStatus.NOTAMMO

            val proj = EntityGaussPulse(worldIn, playerIn, status)
            if (!worldIn.isRemote) worldIn.spawnEntityInWorld(proj)
            val look = playerIn.lookVec
            if (look.xCoord != 0.0 && look.zCoord != 0.0)
                playerIn.knockBack(playerIn, 0.5f, look.xCoord, look.zCoord)
            else
                playerIn.motionY += 0.5
            worldIn.playSound(null, playerIn.posX, playerIn.posY, playerIn.posZ, PsiSoundHandler.cadShoot, SoundCategory.PLAYERS, 1f, 1f)

            if (ammo != null && !playerIn.capabilities.isCreativeMode)
                playerIn.cooldownTracker.setCooldown(this, 50)
        }
        return ActionResult(EnumActionResult.SUCCESS, itemStackIn)
    }

    private fun findAmmo(player: EntityPlayer): ItemStack? {
        if (player.heldItemOffhand?.item == ModItems.gaussBullet) {
            return player.getHeldItem(EnumHand.OFF_HAND)
        } else if (player.heldItemMainhand?.item == ModItems.gaussBullet) {
            return player.getHeldItem(EnumHand.MAIN_HAND)
        } else {
            for (i in 0..player.inventory.sizeInventory - 1) {
                val itemstack = player.inventory.getStackInSlot(i)

                if (itemstack?.item == ModItems.gaussBullet) {
                    return itemstack
                }
            }

            return null
        }
    }

    @SideOnly(Side.CLIENT)
    override fun getItemColor(): IItemColor {
        return IItemColor {
            itemStack, i ->
            if (i == 0)
                pulseColor(0xB87333)
            else if (i == 1)
                ICADColorizer.DEFAULT_SPELL_COLOR
            else 0xFFFFFF
        }
    }

    fun pulseColor(rgb: Int): Int {
        val add = (Math.sin(ClientTickHandler.ticksInGame * 0.2) * 24).toInt()
        val r = (rgb and (0xFF shl 16)) shr 16
        val b = (rgb and (0xFF shl 8)) shr 8
        val g = (rgb and (0xFF shl 0)) shr 0
        return (Math.max(Math.min(r + add, 255), 0) shl 16) or
               (Math.max(Math.min(b + add, 255), 0) shl 8) or
               (Math.max(Math.min(g + add, 255), 0) shl 0)
    }
}